pkg/dns/google/google.go

```diff
@@ -0,0 +1,497 @@
+
+	currentChange := &dnsv1.Change{}
+	var totalChanges int
+	for _, name := range names {
+		c := changesByName[name]
+
+		totalChangesByName := len(c.additions) + len(c.deletions)
+
+		if totalChangesByName > batchSize {
+			g.logger.V(1).Info("Total changes for %s exceeds max batch size of %d, total changes: %d", name,
+				batchSize, totalChangesByName)
+			continue
+		}
+
+		if totalChanges+totalChangesByName > batchSize {
+			totalChanges = 0
+			changes = append(changes, currentChange)
+			currentChange = &dnsv1.Change{}
+		}
+
+		currentChange.Additions = append(currentChange.Additions, c.additions...)
+		currentChange.Deletions = append(currentChange.Deletions, c.deletions...)
+
+		totalChanges += totalChangesByName
+	}
+
+	if totalChanges > 0 {
+		changes = append(changes, currentChange)
+	}
+
+	return changes
+}
+
+// getResourceRecordSets returns the records for a managed zone of the currently configured provider.
+func (g *GoogleDNSProvider) getResourceRecordSets(ctx context.Context, zoneID string) ([]*dnsv1.ResourceRecordSet, error) {
+	var records []*dnsv1.ResourceRecordSet
+
+	f := func(resp *dnsv1.ResourceRecordSetsListResponse) error {
+		records = append(records, resp.Rrsets...)
+		return nil
+	}
+
+	if err := g.resourceRecordSetsClient.List(g.project, zoneID).Pages(ctx, f); err != nil {
+		return nil, err
+	}
+
+	return records, nil
+}
+
+// toResourceRecordSets converts a list of endpoints into `ResourceRecordSet` resources.
+func toResourceRecordSets(allEndpoints []*v1alpha1.Endpoint) []*dnsv1.ResourceRecordSet {
+	var records []*dnsv1.ResourceRecordSet
+
+	// Google DNS requires a record to be created per `dnsName`, so the first thing we need to do is group all the
+	// endpoints with the same dnsName together.
+	endpointMap := make(map[string][]*v1alpha1.Endpoint)
+	for _, ep := range allEndpoints {
+		endpointMap[ep.DNSName] = append(endpointMap[ep.DNSName], ep)
+	}
+
+	for dnsName, endpoints := range endpointMap {
+		// A set of endpoints belonging to the same group(`dnsName`) must always be of the same type, have the same ttl
+		// and contain the same rrdata (weighted or geo), so we can just get that from the first endpoint in the list.
+		ttl := int64(endpoints[0].RecordTTL)
+		recordType := endpoints[0].RecordType
+		_, weighted := endpoints[0].GetProviderSpecificProperty(dns.ProviderSpecificWeight)
+		_, geoCode := endpoints[0].GetProviderSpecificProperty(dns.ProviderSpecificGeoCode)
+
+		record := &dnsv1.ResourceRecordSet{
+			Name: ensureTrailingDot(dnsName),
+			Ttl:  ttl,
+			Type: recordType,
+		}
+		if weighted {
+			record.RoutingPolicy = &dnsv1.RRSetRoutingPolicy{
+				Wrr: &dnsv1.RRSetRoutingPolicyWrrPolicy{},
+			}
+		} else if geoCode {
+			record.RoutingPolicy = &dnsv1.RRSetRoutingPolicy{
+				Geo: &dnsv1.RRSetRoutingPolicyGeoPolicy{},
+			}
+		}
+
+		for _, ep := range endpoints {
+			geoCodeProp, _ := ep.GetProviderSpecificProperty(dns.ProviderSpecificGeoCode)
+			if geoCodeProp.Value != "*" {
```

Comment:
@R-Lawton Something is off with this logic. In the case where don't specify a geo-code (weighted only) we will only ever have the single wildcard entry and removing it causes the dns chain of records to break as it's going to be missing any records for the gateway lb.

I had a go and changing it [here](https://github.com/Kuadrant/multicluster-gateway-controller/commit/55ff95795d94099bca4bc46c1a63bc83af829dbd#diff-eea59a6fb9851390f6210a03f57ba88ea1628c32d20d526998600135a1642bed), feels a bit hacky, but might get us to a point of being able to merge this and look into a better solution after.